File: src/brs/web/api/http/handler/GenerateDeeplinkQR.java
package brs.web.api.http.handler;

import brs.deeplink.DeeplinkGenerator;
import brs.web.api.http.ApiServlet.HttpRequestHandler;
import brs.util.Convert;
import brs.util.StringUtils;
import brs.web.api.http.common.LegacyDocTag;
import com.google.zxing.WriterException;
import org.eclipse.jetty.http.HttpStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.imageio.ImageIO;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.awt.image.BufferedImage;
import java.io.IOException;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;

public class GenerateDeeplinkQR extends HttpRequestHandler {

  private final Logger logger = LoggerFactory.getLogger(GenerateDeeplinkQR.class);
  public static final GenerateDeeplinkQR instance = new GenerateDeeplinkQR();

  private GenerateDeeplinkQR() {
    super(new LegacyDocTag[]{LegacyDocTag.UTILS}, ACTION_PARAMETER, PAYLOAD_PARAMETER);
  }

  @Override
  public void processRequest(HttpServletRequest req, HttpServletResponse resp) {
    try {

      final String action = Convert.emptyToNull(req.getParameter(ACTION_PARAMETER));
      final String payload = Convert.emptyToNull(req.getParameter(PAYLOAD_PARAMETER));

      if (StringUtils.isEmpty(action) && !StringUtils.isEmpty(payload)) {
        addErrorMessage(resp, PAYLOAD_WITHOUT_ACTION);
        return;
      }

      DeeplinkGenerator deeplinkGenerator = new DeeplinkGenerator();
      try {
        final BufferedImage qrImage = deeplinkGenerator.generateDeepLinkQrCode(action, payload);
        resp.setContentType("image/jpeg");
        ImageIO.write(qrImage, "jpg", resp.getOutputStream());
        resp.getOutputStream().close();
      } catch (IllegalArgumentException e) {
        logger.error("Problem with arguments", e);
        addErrorMessage(resp, incorrect("arguments", e.getMessage()));
      }
    } catch (WriterException | IOException e) {
      logger.error("Could not generate Deeplink QR code", e);
      resp.setStatus(HttpStatus.INTERNAL_SERVER_ERROR_500);
    }
  }
}

File: src/main/kotlin/brs/api/http/GenerateDeeplinkQR.kt
package brs.api.http

import brs.api.http.APIServlet.HttpRequestHandler
import brs.api.http.JSONResponses.MISSING_DOMAIN
import brs.api.http.JSONResponses.PAYLOAD_WITHOUT_ACTION
import brs.api.http.JSONResponses.incorrect
import brs.api.http.common.Parameters.ACTION_PARAMETER
import brs.api.http.common.Parameters.DOMAIN_PARAMETER
import brs.api.http.common.Parameters.PAYLOAD_PARAMETER
import brs.entity.DependencyProvider
import brs.util.convert.emptyToNull
import brs.util.logging.safeDebug
import com.google.zxing.WriterException
import org.eclipse.jetty.http.HttpStatus
import org.slf4j.LoggerFactory
import java.io.IOException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

internal class GenerateDeeplinkQR(private val dp: DependencyProvider) : HttpRequestHandler(arrayOf(APITag.UTILS), DOMAIN_PARAMETER, ACTION_PARAMETER, PAYLOAD_PARAMETER) {
    override fun processRequest(request: HttpServletRequest, resp: HttpServletResponse) {
        try {
            val domain = request.getParameter(DOMAIN_PARAMETER).emptyToNull()
            if (domain.isNullOrEmpty()) {
                addErrorMessage(resp, MISSING_DOMAIN)
                return
            }

            val action = request.getParameter(ACTION_PARAMETER).emptyToNull()
            val payload = request.getParameter(PAYLOAD_PARAMETER).emptyToNull()

            if (action.isNullOrEmpty() && !payload.isNullOrEmpty()) {
                addErrorMessage(resp, PAYLOAD_WITHOUT_ACTION)
                return
            }

            try {
                val url = dp.deeplinkGeneratorService.generateDeepLink(domain, action, payload)
                resp.contentType = "image/png"
                dp.qrGeneratorService.writeQRCodeToStream(url, "png", resp.outputStream)
                resp.outputStream.close()
            } catch (e: IllegalArgumentException) {
                logger.safeDebug(e) { "Problem with arguments" }
                addErrorMessage(resp, incorrect("arguments", e.message))
            }
        } catch (e: WriterException) {
            logger.safeDebug(e) { "Could not generate Deeplink QR code" }
            resp.status = HttpStatus.INTERNAL_SERVER_ERROR_500
        } catch (e: IOException) {
            logger.safeDebug(e) { "Could not generate Deeplink QR code" }
            resp.status = HttpStatus.INTERNAL_SERVER_ERROR_500
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(GenerateDeeplinkQR::class.java)
    }
}
